File: sandwichclub/src/main/java/de/pacheco/sandwichclub/SandwichDetailActivity.java
package de.pacheco.sandwichclub;

import de.pacheco.sandwichclub.model.Sandwich;
import de.pacheco.sandwichclub.utils.JsonUtils;

import android.content.Intent;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.squareup.picasso.Picasso;

public class SandwichDetailActivity extends AppCompatActivity {

    public static final String EXTRA_POSITION = "extra_position";
    private static final int DEFAULT_POSITION = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_detail);
        ImageView ingredientsIv = findViewById(R.id.image_iv);
        Intent intent = getIntent();
        if (intent == null) {
            closeOnError();
        }
        @SuppressWarnings("ConstantConditions") int position = intent.getIntExtra(EXTRA_POSITION,
                DEFAULT_POSITION);
        if (position == DEFAULT_POSITION) {
            // EXTRA_POSITION not found in intent
            closeOnError();
            return;
        }
        String[] sandwiches = getResources().getStringArray(R.array.sandwich_details);
        String json = sandwiches[position];
        Sandwich sandwich = JsonUtils.parseSandwichJson(json);
        if (sandwich == null) {
            // Sandwich data unavailable
            closeOnError();
            return;
        }
        populateUI(sandwich);
        Picasso.get()
                .load(sandwich.getImage())
                .into(ingredientsIv);
        setTitle(sandwich.getMainName());
    }

    private void closeOnError() {
        Toast.makeText(this, R.string.detail_error_message, Toast.LENGTH_SHORT).show();
        finish();
    }

    private void populateUI(Sandwich sandwich) {
        setText(R.id.also_known_tv, sandwich.getAlsoKnownAs());
        setText(R.id.ingredients_tv, sandwich.getIngredients());
        setText(R.id.origin_tv, sandwich.getPlaceOfOrigin());
        setText(R.id.description_tv, sandwich.getDescription());
    }

    private void setText(int id, Object text) {
        TextView view = findViewById(id);
        String finalText = stripText(text.toString()) + System.lineSeparator();
        view.setText(finalText);
    }

    private String stripText(String text) {
        text = text.startsWith("[") ? text.substring(1) : text;
        text = text.endsWith("]") ? text.substring(0, text.length() - 1) : text;
        return text;
    }
}

File: sandwichclub/src/main/java/de/pacheco/sandwichclub/SandwichDetailActivity.kt
package de.pacheco.sandwichclub

import android.os.Bundle
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.squareup.picasso.Picasso
import de.pacheco.sandwichclub.model.Sandwich
import de.pacheco.sandwichclub.utils.parseSandwichJson

class SandwichDetailActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_detail)
        val ingredientsIv = findViewById<ImageView>(R.id.image_iv)
        val intent = intent
        if (intent == null) {
            closeOnError()
        }
        val position = intent!!.getIntExtra(EXTRA_POSITION, DEFAULT_POSITION)
        if (position == DEFAULT_POSITION) {
            // EXTRA_POSITION not found in intent
            closeOnError()
            return
        }
        val sandwiches = resources.getStringArray(R.array.sandwich_details)
        val json = sandwiches[position]
        val sandwich = parseSandwichJson(json)
        populateUI(sandwich)
        Picasso.get()
                .load(sandwich.image)
                .into(ingredientsIv)
        title = sandwich.mainName
    }

    private fun closeOnError() {
        Toast.makeText(this, R.string.detail_error_message, Toast.LENGTH_SHORT).show()
        finish()
    }

    private fun populateUI(sandwich: Sandwich) {
        setText(R.id.also_known_tv, sandwich.alsoKnownAs)
        setText(R.id.ingredients_tv, sandwich.ingredients)
        setText(R.id.origin_tv, sandwich.placeOfOrigin)
        setText(R.id.description_tv, sandwich.description)
    }

    private fun setText(id: Int, text: Any) {
        val view = findViewById<TextView>(id)
        val finalText = stripText(text.toString()) + System.lineSeparator()
        view.text = finalText
    }

    private fun stripText(originalText: String): String {
        var text = originalText
        text = if (text.startsWith("[")) text.substring(1) else text
        text = if (text.endsWith("]")) text.substring(0, text.length - 1) else text
        return text
    }

    companion object {
        const val EXTRA_POSITION = "extra_position"
        private const val DEFAULT_POSITION = -1
    }
}